### PR TITLE
watchdog: Touch Watchdog before executing most event loop callbacks to avoid spurious misses when there are many queued callbacks.

### PR DESCRIPTION
**Pull request by antoniovicente**
Commit Message: watchdog: Touch Watchdog before executing most event loop callbacks to avoid spurious misses when there are many queued callbacks.
Additional Description:
This change assumes that the primary purposes of the watchdog is to terminate the proxy if worker threads are deadlocked, and aid in the debugging of long-running operations that end up scheduled in worker threads through monitoring for miss/megamiss events and auxiliary watchdog actions like the recently added watchdog profile action.
Risk Level: medium. more frequent calls to the watchdog could have visible performance implications; mitigated by #13103
Testing: unit
Docs Changes: n/a
Release Notes: added
Runtime guard: n/a
Fixes: #11391

### CODE REVIEW

**Review by KBaichoo — COMMENTED** [review]
Nice, thanks for implementing this! It will make the watchdog system more useful.

Curious if have any data on the performance impact

File: include/envoy/server/watchdog.h
Comment (antoniovicente) [reply include/envoy/server/watchdog.h:27]: It depends on how real instances of this interface are created.  WatchdogImpl is always registered with a dispatcher.  Manually touching the watchdog continues to be a valid.

File: source/common/event/dispatcher_impl.h
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.h:150]: I think it can happen in tests, I can see how common it is and try to restrict this to a single watchdog per dispatcher.

**Comment by antoniovicente** [comment]
> Nice, thanks for implementing this! It will make the watchdog system more useful.
> 
> Curious if have any data on the performance impact

I'm not fully up to speed on available OSS performance tests, so no - I don't know about the full performance implications of this change with or without #13103

File: source/common/event/dispatcher_impl.h
Comment (KBaichoo) [inline source/common/event/dispatcher_impl.h:97]: nit: maybe add a comment for this class

File: source/common/event/dispatcher_impl.h
Comment (KBaichoo) [reply source/common/event/dispatcher_impl.h:150]: Ah. It probably doesn't happen in production code IIRC. 

Up to you if you want to squeeze out some additional performance by just having a single pointer here at the cost of unit tests being a bit more messy.

**Review by KBaichoo — COMMENTED** [review]
lgtm besides a few nits

I'll defer to senior maintainers regarding any performance tests we have / if we want to measure the impact of this

File: source/common/event/dispatcher_impl.h
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.h]: Changed to support just 1.  The only affected test was the one I added for multiple watchdogs registered with the same dispatcher.

File: source/common/event/dispatcher_impl.h
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.h:99]: Done.

**Comment by antoniovicente** [comment]
Back from vacation, waiting until #13103 is merged in.

File: source/common/event/dispatcher_impl.h
Comment (KBaichoo) [inline source/common/event/dispatcher_impl.h:125]: used to touch the watchdog*

**Review by KBaichoo — DISMISSED** [review]
LGTM

File: source/common/event/dispatcher_impl.h
Comment (antoniovicente) [reply source/common/event/dispatcher_impl.h]: done.

**Review by KBaichoo — DISMISSED** [review]

**Comment by antoniovicente** [comment]
@envoyproxy/senior-maintainers for final review.

**Review by mattklein123 — COMMENTED** [review]
Thanks for working on this.

qq: did we consider just doing the touch as part of the prepare callback that runs before each event loop iteration? Would that be good enough as opposed to touching before every individual event? I know you made touching faster but there is still cost so just wondering if we can get most of the benefit by just doing it in the prepare callback (or maybe both pre-post callbacks for each iteration)?

/wait-any

**Comment by antoniovicente** [comment]
> Thanks for working on this.
> 
> qq: did we consider just doing the touch as part of the prepare callback that runs before each event loop iteration? Would that be good enough as opposed to touching before every individual event? I know you made touching faster but there is still cost so just wondering if we can get most of the benefit by just doing it in the prepare callback (or maybe both pre-post callbacks for each iteration)?
> 
> /wait-any

I don't think touching on prepare is enough.

**Comment by mattklein123** [comment]
> I don't think touching on prepare is enough.

OK fair enough!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**
